lib/matplotlib/axis.py

```diff
@@ -1626,14 +1632,28 @@ def set_ticklabels(self, ticklabels, *, minor=False, **kwargs):
         """
         ticklabels = [t.get_text() if hasattr(t, 'get_text') else t
                       for t in ticklabels]
+        locator = (self.get_minor_locator() if minor
+                   else self.get_major_locator())
+        if isinstance(locator, mticker.FixedLocator):
+            tickd = {loc: lab for loc, lab in zip(locator.locs, ticklabels)}
+            func = functools.partial(self._format_with_dict, tickd)
+            formatter = mticker.FuncFormatter(func)
+        else:
+            formatter = mticker.FixedFormatter(ticklabels)
+
         if minor:
-            self.set_minor_formatter(mticker.FixedFormatter(ticklabels))
-            ticks = self.get_minor_ticks()
+            self.set_minor_formatter(formatter)
+            locs = self.get_minorticklocs()
+            ticks = self.get_minor_ticks(len(locs))
         else:
-            self.set_major_formatter(mticker.FixedFormatter(ticklabels))
-            ticks = self.get_major_ticks()
+            self.set_major_formatter(formatter)
+            locs = self.get_majorticklocs()
+            ticks = self.get_major_ticks(len(locs))
+
         ret = []
-        for tick_label, tick in zip(ticklabels, ticks):
+        for pos, (loc, tick) in enumerate(zip(locs, ticks)):
+            tick.update_position(loc)
+            tick_label = formatter(tick._loc, pos)
```

Comment:
is `tick._loc` just `loc`? or not?

Reply:
At this point they should be identical; but yes, it might look better and be slightly more efficient with `loc`.